Document Jinja filter schema usage

dwyeradam opened the pull request:
### What is the context of this PR?
Document Jinja filters used in schemas
Also:
Removes unused `format_currency_for_input` filter
Renames `format_household_member_name` methods to match names used to call them
Adds a unit test for `test_format_household_summary` which was missing

### How to review 
Read the new `jinja-filters.md` file and help me make it less boring/more understandable.
Also check the couple of minor out of scope changes haven't had a negative effect.

### Checklist

* [ ] New static content marked up for translation
* [ ] Newly defined schema content included in eq-translations repo

dwyeradam committed:
Document Jinja filter schema usage

Additionally:
Removes unused format_currency_for_input filter
Renames format_household_member_name methods to match names used to call them
Adds a unit test for test_format_household_summary which was missing
Change param names for format_address_list to be more generic

MebinAbraham commented on doc/jinja-filters.md:
```diff
@@ -0,0 +1,232 @@
+
+##### Context:
+No real current usage but a single question in Census I believe only for test purposes.
+`The number of visitors staying in the household on {{ get_current_date() }}`
```
Just for consistency, can we add break tags before these so they are not on the same line?

ajmaddaford replied:
It's cleaner to put the fixed width content in a ``` block, something like:
```
The number of visitors staying in the household on {{ get_current_date() }}
```

ajmaddaford commented on app/templating/template_renderer.py:
```diff
@@ -15,8 +15,8 @@ def __init__(self):
-        env.filters['format_household_name'] = filters.format_household_member_name
-        env.filters['format_household_name_possessive'] = filters.format_household_member_name_possessive
+        env.filters['format_household_name'] = filters.format_household_name
+        env.filters['format_household_name_possessive'] = filters.format_household_name_possessive
```
👍

ajmaddaford commented on doc/jinja-filters.md:
```diff
@@ -0,0 +1,248 @@
+# Jinja Filters Used by Schemas
+
+###### Date 14/12/2018
```
I don't think we need the date here as we'll have the date it merges to master in source control.

ajmaddaford commented on doc/jinja-filters.md:
```diff
@@ -0,0 +1,248 @@
+##### Context:
+Most commonly used to display a previous answer's numeric value in a following question.
+<br />
+`Of the <em>{{answers['total-number-employees']|format_number}}</em> total employees employed on 14 December 2018, how many male and female employees worked the following hours?`
```
It's not clear what does `format_number` does in this example. It was only after looking into `format_decimal` that I realised it will add comma separators to larger numbers.

dwyeradam replied:
I'll add some additional text but I also think the Output examples will really help with understanding all the filters especially these smaller ones.

ajmaddaford commented on doc/jinja-filters.md:
```diff
@@ -0,0 +1,248 @@
+##### Context:
+Used to display a full address for use in a description.
+<br />
+`"description": "{{ format_address_list ([answers['address-line1'], answers['address-line2'], answers['locality'], answers['town-name'], answers['postcode']], [metadata['address_line1'], metadata['address_line2'], metadata['locality'], metadata['town_name'], metadata['postcode']]) }}"`
```
Some strings contain the json element e.g. description, and others do not. As it doesn't seem to add anything I think we should remove them from all the examples.

ajmaddaford commented on doc/jinja-filters.md:
```diff
@@ -0,0 +1,248 @@
+
+
+## format_date
+Formats a date string, returning "d MMMM YYYY", "MMMM YYYY" or "YYYY" depending on value passed. 
```
I think this needs to be clearer. What values can be passed, and what will the output be for those values?

ajmaddaford commented on doc/jinja-filters.md:
```diff
@@ -0,0 +1,248 @@
+
+
+## format_conditional_date
+This is to be used when passing multiple dates as multiple arguments. It will return the first valid date formatted.
```
Possibly not for this pull request, but I wonder now that we have `first_non_empty_item` whether we could combine it with `format_date` so we don't need this filter, something like:
```
{{ first_non_empty_item(answers['period-from'], metadata['ref_p_start_date']) | format_date }}
```

dwyeradam replied:
I agree and thought exactly the same. I'd like to check with Mebin again as there was a reason he didn't do it at the time of writing this new filter.

MebinAbraham replied:
So `format_conditional_date` reuses the same private method that `first_non_empty_item` uses `_get_first_non_empty_item`.

Having it as two filter just mean't I didn't have to essentially duplicate
``` python
	date_format = 'd MMMM YYYY'
    if value and re.match(r'\d{4}-\d{2}$', value):
        date_format = 'MMMM YYYY'
    if value and re.match(r'\d{4}$', value):
        date_format = 'YYYY'
```
to check whether what it got passed in regex matched any one of the different formats we support before passing it into the `format_date` filter and then just returning the value if it wasn't a date. I guess you could also try: except to see if it would parse as a date. 🤷‍♂️ 

Just wasn't sure if this would have be very nice.

ajmaddaford commented on doc/jinja-filters.md:
```diff
@@ -0,0 +1,248 @@
+## calculate_offset_from_weekday_in_last_whole_week
+Offsets a date from a particular day of the week in the previous week. 
+Intended to be used to generate reference date ranges around a particular date
+The offsets will always be based on one day of the week in the previous Mon-Sun
```
It's still not clear what this does, more examples with output could help.

ajmaddaford reviewed (changes requested):
This definitely helps highlight what filters we currently have. For each example I think it would be helpful to also document the output. Also, for some of the filters that can be passed different values, it would be useful to have more than one example.

ajmaddaford commented on doc/jinja-filters.md:
```diff
@@ -0,0 +1,438 @@
+
+## format_number
+Returns the given number with the appropriate thousands grouping and decimal separator as determined by the locale.
+Uses Babel's number.py and format_decimal.
```
We could link to the Babel documentation for this method, something like:

Uses Babel's [format_decimal](http://babel.pocoo.org/en/latest/api/numbers.html#babel.numbers.format_decimal) method.

ajmaddaford commented on doc/jinja-filters.md:
````diff
@@ -0,0 +1,438 @@
+```
+
+## format_currency
+Returns a formatted currency value. Uses Babel's number.py and format_currency.
````
Same as comment above about linking to the Babel documentation.

ajmaddaford commented on doc/jinja-filters.md:
```diff
@@ -0,0 +1,438 @@
+
+##### Parameters: 
+- user_entered_address: List of address values from answers
+- metadata_address: List of address values from metadata
```
This method could do with being updated to remove the specifics about user vs metadata address. It looks like it just takes two addresses and outputs the first if it's not empty, else the second.

dwyeradam replied:
I really wanted to do this so I will. I was worried about moving the scope too much but I really hate it so we've got two votes to change it as part of this.

ajmaddaford commented on doc/jinja-filters.md:
```diff
@@ -0,0 +1,438 @@
+
+
+## get_current_date
+Returns the current Date.
```
Date -> date

ajmaddaford commented on doc/jinja-filters.md:
```diff
@@ -0,0 +1,438 @@
+- N/A
+
+##### Context:
+No real current usage but a single question in Census I believe only for test purposes.
```
We should be more definitive here and remove "I believe only ".

ajmaddaford commented on doc/jinja-filters.md:
```diff
@@ -0,0 +1,438 @@
+The output will be of the form "d MMMM YYYY", "MMMM YYYY" or "YYYY" depending on value passed.
+
+##### Parameters: 
+- value: Value representing a datetime.
```
This method looks like it will only do something if it's passed a string (if it's passed anything else it just returns the value). It might be clearer to move the acceptable string formats from the description above to here.

bitdivision commented on doc/jinja-filters.md:
```diff
@@ -0,0 +1,438 @@
+("2018-08-10", {"years": 1}, "FR")  "2019-08-03"  # Friday outputs previous Friday + 1 year
+
+("2018-08-05", {}, "SU")            "2018-07-29"  # Sunday outputs previous Sunday (Must be a full Sunday)
+("2018-08-05", {"weeks": 1}, "SU")  "2018-08-05"  # Previous sunday with +1 month offset, back to input
```
Should this be `+1 week`?

dwyeradam replied:
Yes but since I copied this from your test_jinja_filters test I blame you. I'll update both now.

ajmaddaford commented on doc/jinja-filters.md:
```diff
@@ -0,0 +1,438 @@
+Returns a date in a defined format. In most cases in order to include the day of the week.
+
+##### Parameters: 
+- value: Value representing a datetime.
```
Should mention the string formats mentioned in the comment about `format_date` above.

ajmaddaford commented on doc/jinja-filters.md:
```diff
@@ -0,0 +1,438 @@
+
+
+## format_date_custom
+Returns a date in a defined format. In most cases in order to include the day of the week.
```
Should mention the Babel [format_date](http://babel.pocoo.org/en/latest/api/dates.html#babel.dates.format_date) method that's used here.

ajmaddaford commented on doc/jinja-filters.md:
```diff
@@ -0,0 +1,438 @@
+Input:                              Output:
+("2016-01-12", "2016-02-12")        "<span class='date'>12 January 2016</span>"
+("2017-12-23", None)                "<span class='date'>23 December 2017</span>"
+(None, "2017-12-24")                "<span class='date'>24 December 2017</span>"
```
The examples would be clearer if the first and second dates were consistent.

ajmaddaford commented on doc/jinja-filters.md:
```diff
@@ -0,0 +1,438 @@
+- *dates: Multiple date values
+
+##### Context:
+If a user decides to report for different dates then following questions should use the user entered date. Otherwise continue asking for the metadata period.
```
This is very specific to how we use this filter rather than what the filter does.

dwyeradam replied:
That's what context means? `"the situation within which something exists or happens"`. You're right though I should be more generic with how to use it.

ajmaddaford commented on doc/jinja-filters.md:
```diff
@@ -0,0 +1,438 @@
+("2018-08-10", {"years": 1}, "FR")  "2019-08-03"  # Friday outputs previous Friday + 1 year
+
+("2018-08-05", {}, "SU")            "2018-07-29"  # Sunday outputs previous Sunday (Must be a full Sunday)
+("2018-08-05", {"weeks": 1}, "SU")  "2018-08-05"  # Previous sunday with +1 month offset, back to input
```
This is a +1 **week** offset

ajmaddaford commented on doc/jinja-filters.md:
```diff
@@ -0,0 +1,438 @@
+("2018-08-10", {}, "FR")            "2018-08-03"  # Friday outputs previous Friday
+("2018-08-10", {"years": 1}, "FR")  "2019-08-03"  # Friday outputs previous Friday + 1 year
+
+("2018-08-05", {}, "SU")            "2018-07-29"  # Sunday outputs previous Sunday (Must be a full Sunday)
```
What does must be a full Sunday mean?

bitdivision replied:
I believe it was intended to show that the current Sunday (e.g. today) would never be output.

dwyeradam replied:
As Richard says. I've changed it to say must be a full week which is clearer.

ajmaddaford commented on doc/jinja-filters.md:
````diff
@@ -0,0 +1,438 @@
+##### Examples:
+```
+Input:                  Output:
+[['item 1', 'item 2']]  "<ul><li>item 1</li><li>item 2</li></ul>"
````
Is the list within a list correct?

dwyeradam replied:
Yes. It's one of the tests. Will try and clarify it's use if I actually understand it.

ajmaddaford replied:
I've had a look at why this is and it's because the list of checkbox answers are being appended  to another list at https://github.com/ONSdigital/eq-survey-runner/blob/master/app/templating/schema_context.py#L116. It does this as it also deals with multiple answers in repeating groups where the answer value could not be a list. We should only create a list if the answer being returned is not already a list.

dwyeradam replied:
Added card https://trello.com/c/DrZyF3sP to address this.

ajmaddaford commented on doc/jinja-filters.md:
```diff
@@ -0,0 +1,438 @@
+Ouput: "<ul>
+            <li>John Smith</li>
+            <li>Jane Smith</li>
+            <li>first3 last3</li>
```
I think this line needs to go

ajmaddaford commented on doc/jinja-filters.md:
```diff
@@ -0,0 +1,438 @@
+
+Ouput: "<ul>
+            <li>51 Testing Gardens, Bristol, BS9 1AW</li>
+       </ul>"
```
Where do we use this filter for an address?

dwyeradam replied:
We don't. I just wanted to show an example where the delimiter is changed and address seemed the best fit for that. I could remove it or have a much simpler example of Surname, First Name. Is this PR just about how it's currently used or how it could be used? Have I just opened myself up for even more changes?

ajmaddaford replied:
I'd say it's about how it's currently used, as we want it to drive the schema changes for piping.

MebinAbraham commented on doc/jinja-filters.md:
```diff
@@ -0,0 +1,438 @@
+Returns the first non empty value in a list of values.
+
+##### Parameters: 
+- *items: List of items where some of them could have no value.
```
Not sure but `List of items` to me sounds like 
``` python 
first_non_empty_item([1,'',3])
# returns [1, '', 3]
```
instead of
``` python
first_non_empty_item(1, '', 3)
# returns 1
```
Maybe `1 or more arguments`?
🤷‍♂️

MebinAbraham commented on doc/jinja-filters.md:
```diff
@@ -0,0 +1,438 @@
+- *items: List of items where some of them could have no value.
+
+##### Context:
+Implemented to replace trad_as_or_runame hard coding. 
```
`trad_as_or_ru_name`

dwyeradam commented:
@ajmaddaford, @MebinAbraham I should have addressed the comments now so ready for re-review. @ajmaddaford `format_unordered_list` is an odd one. That's the format that is passed into it. We may have to look at how we store the answer here. For now I've tried to be clearer on what format the filter is expecting.

MebinAbraham commented on app/jinja_filters.py:
```diff
@@ -42,35 +42,26 @@ def get_currency_symbol(currency='GBP'):
-        :return: the value of the address to be piped
+        :param (list) first_address
+        :param (list) second_address
+        :return: first address is values present else second address
```
`if values`

MebinAbraham commented on doc/jinja-filters.md:
```diff
@@ -0,0 +1,445 @@
+
+
+## format_household_summary
+A merge of format_household_name and format_unordered_list in order to return a list of full names in "ul" and "li' tags 
```
If you got some free time on your hands, feel free to highlight any variable/function names in sentences. 

A merge of `format_household_name` and `format_unordered_list` in order to return a list of full names in "ul" and "li' tags

dwyeradam replied:
No I need you to point them all out.

dwyeradam replied:
I've probably got them all. In the end not that many.
